src/istio/mixerclient/client_impl.cc

```diff
@@ -170,82 +191,136 @@ CancelFunc MixerClientImpl::Check(CheckContextSharedPtr &context,
-
-                     if (!context->policyCacheHit()) {
-                       context->updatePolicyCache(status, *context->response());
-
-                       if (context->policyStatus().ok()) {
-                         ++total_remote_check_accepts_;
-                       } else {
-                         ++total_remote_check_denies_;
-                       }
-                     }
-
-                     if (context->quotaCheckRequired()) {
-                       context->updateQuotaCache(status, *context->response());
-
-                       if (context->quotaStatus().ok()) {
-                         ++total_remote_quota_accepts_;
-                       } else {
-                         ++total_remote_quota_denies_;
-                       }
-                     }
-
-                     //
-                     // Determine final status for Filter::completeCheck(). This
-                     // will send an error response to the downstream client if
-                     // the final status is not Status::OK
-                     //
-
-                     if (result != TransportResult::SUCCESS) {
-                       if (context->networkFailOpen()) {
-                         context->setFinalStatus(Status::OK);
-                       } else {
-                         context->setFinalStatus(status);
-                       }
-                     } else if (!context->quotaCheckRequired()) {
-                       context->setFinalStatus(context->policyStatus());
-                     } else if (!context->policyStatus().ok()) {
-                       context->setFinalStatus(context->policyStatus());
-                     } else {
-                       context->setFinalStatus(context->quotaStatus());
-                     }
-
-                     if (on_done) {
-                       on_done(*context);
-                     }
-
-                     if (utils::InvalidDictionaryStatus(status)) {
-                       compressor_.ShrinkGlobalDictionary();
-                     }
-                   });
+  if (remote_quota_prefetch) {
+    ++total_remote_quota_prefetch_calls_;
+  }
+
+  RemoteCheck(context, transport ? transport : options_.env.check_transport,
+              remote_quota_prefetch ? nullptr : on_done);
+}
+
+void MixerClientImpl::RemoteCheck(CheckContextSharedPtr context,
+                                  const TransportCheckFunc &transport,
+                                  const CheckDoneFunc &on_done) {
+  //
+  // This lambda and any lambdas it creates for retry will inc the ref count
+  // on the CheckContext shared pointer.
+  //
+  // The other captures (this/MixerClientImpl, transport func, and on_done func
+  // are assumed to be long-lived objects that exist until shutdown.
+  //
+  CancelFunc cancel_func = transport(
+      context->request(), context->response(),
+      [this, context, transport, on_done](const Status &status) {
+        context->resetCancel();
```

Comment:
I am confused by this cancel stuff.
- When this lambda is invoked, the `resetCancel` invalid future cancel attempt
- this lambda doesn't always succeed. A retry may occurs.

In case there is a retr, does the `retry` retore the `cancel`? If so, where is it? If not, is it a big issue?

Reply:
I'll try to explain....

Transactions are cancelled when the downstream client goes away and when the overall transaction times out.

In the original code there was a cancel lambda that is invoked when the transaction is cancelled.  This cancel lambda tells the grpc library to cancel any outstanding grpc request sent to the mixer server.   If a request is not outstanding, then the cancel lambda will be null and won't be called.

In the new code, it's also possible for requests to be in a retry wait when the transaction is cancelled.   In this case we don't tell the grpc library to cancel the request because no request has been passed to it.  Instead we tell envoy to cancel the retry timer.

In check_context.h you'll find this:

```
  void cancel() {
    if (cancel_func_) {
      MIXER_DEBUG("Cancelling check call");
      cancel_func_();
      cancel_func_ = nullptr;
    }

    if (retry_timer_) {
      MIXER_DEBUG("Cancelling retry");
      retry_timer_->Stop();
      retry_timer_ = nullptr;
    }
  }
```

Reply:
And yes, once the retry lambda is invoked, it will restore the cancel lambda assuming it can successfully pass a request to the grpc library